source/base/quadrature.cc

```diff
@@ -57,6 +57,16 @@ Quadrature<dim>::initialize(const std::vector<Point<dim>> &p,
 
 
 
+template <int dim>
+void
+Quadrature<dim>::push_back(const Point<dim> &point, const double weight)
+{
+  quadrature_points.push_back(point);
+  weights.push_back(weight);
+}
```

Comment:
Should I also set `is_tensor_product_flag=false` here?

Reply:
This problem already exist (if we really try to break something)
If we:
1. Create a tensor product quadrature.
2. Call Quadrature<dim>::initialize with something that isn't a tensor product.

we have a non-tensor product quadrature having `is_tensor_product_flag=true`.  :(

The easy solution would be to just set `is_tensor_product_flag=false` in 
`Quadrature<dim>::initialize`. 

A more cumbersome solution (more along @drwells previous suggestion):

Since 
1. `push_back` and `initialize` doesn't make much sense for a tensor product quadrature.
2. Not every quadrature "is a" tensor product.
One could imagine the following
```c++

template<int dim>
class Quadrature : Subscriptor{
	
	// Same as before, but 
	// 1. initialize(..) removed.
	// 2. tensor_basis removed
	
	// 3. Changed to virtual
	virtual bool is_tensor_product(){ return false;} 
		
	// 4. Throws exception
	virtual const std::array<Quadrature<1>, dim> & get_tensor_basis() const;
};

template<int dim>
class TensorProductQuadrature : public Quadrature<dim> {
	// Constructors creating tensor products moved here.

        virtual bool is_tensor_product(){ return true;} 
	
	// Overloaded.
	const std::array<Quadrature<1>, dim> & get_tensor_basis() const override;
};

template<int dim>
class ExtendableQuadrature : Quadrature<dim>{
	
       // These only exist here
	void initialize(..);
	void push_back(..);
};

template<int dim>
class ImmersedSurfaceQuadrature : public Quadrature<dim>{

// This class no longer has push_back and initialize taking only points and
// weights. So we can't end up in a state with normals having a different size.
// Also, shorter implementation here since we can still derive from Quadrature.

};

```

Reply:
I like this class hierarchy. In fact, you don't even need the `is_tensor_product()` member function -- you can discover whether something is a tensor product quadrature using a `dynamic_cast`: i.e., the property is already encoded in the class hierarchy rather than via virtual member functions.

Reply:
Since `initialize` is removed from `Quadrature` here, this is not really backwards compatible.
On the other hand, the number of users who create custom quadratures is probably small.

Reply:
Yes, I agree. Is that function used anywhere in the library?

Reply:
I don't think so. If I remove it the library still compiles. I didn't run the whole test-suit but searching for
`[q,Q][\w]+\.initialize\(`
finds no occurrence of it.

Reply:
OK, that's a good indication that it's indeed unused. Thanks for trying!